### PR TITLE
Development

### PR DESCRIPTION
**Pull request by Blazemacroz**

### CODE REVIEW

File: api/products.js [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    } catch (error) {
+      console.error(error);
+      next(error)
+    }
+})
+//get all the products 
+router.get("/getProducts", async (req, res, next)=> {
+  try{
+    const products = getAllProducts();
+    res.send("getting all products",products)
```
Comment (parisyee) [inline api/products.js:37]: you don't need this string

File: api/products.js [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+  }catch(error){
+    next(error)
+  }
+})
+
+//delete product(only admin can delete products)
+router.delete("/:productId", async (req, res, next)=>{
```
Comment (parisyee) [inline api/products.js:44]: how are you preventing non admins from deleting this?

File: api/reviews.js [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+const router = express.Router();
+const {
+    createReview,
+    getReviewById,
+    getReviewByProduct
+} = require('../db');
+
+router.get('/', async (req, res, next) => {
+    try {
+        const reviews = await getReviewByProduct();
```
Comment (parisyee) [inline api/reviews.js:11]: Where is the reference to the product?

File: api/utils.js [hunk 4]
```diff
@@ -0,0 +1,16 @@
+async function requireUser(req, res, next) {
+  if (!req.user) {
+    // console.log(req.user);
+    res.send({
```
Comment (parisyee) [inline api/utils.js:4]: this should be a 403 status code I believe "not authorized"

File: app.js [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+const bodyParser = require("body-parser");
+
+app.use(bodyParser.json());
+app.use("/api", apiRouter);
+app.use(morgan("dev"));
+app.use(cors());
+
+app.use((req, res) => {
+  res.status(404).send({
+    message: "REQUEST FAILED 404",
```
Comment (parisyee) [inline app.js:16]: message should be "not found"

File: db/products.js [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+        FROM products;
+        `)
+        return products
+    }catch(error){
+        console.error(error)
+    }
+}
+
+
+async function getProductsById(id) {
```
Comment (parisyee) [inline db/products.js:43]: method name should be singular if you're just finding one.

File: db/reviews.js [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+      `, [rating, comment, review_date]);
+      console.log({ rating, comment, review_date });
+      if (!review) {
+      throw Error
+    } else {
+        return review
+    }
+    } catch (error) {
+        console.log(error)
+    }
```
Comment (parisyee) [inline db/reviews.js:19]: Don't these need to be associated with products and users? Where's the reference to `productId` or `userId`?

**Review by parisyee — CHANGES_REQUESTED** [review]
This is a little concerning. You all should be much further along at this point and the code I'm seeing does not appear to be complete. Also, please make sure all your code is well formatted. Poorly formatted code will be disregarded by potential employers, and it is also difficult for reviewers and other developers to read.

File: .gitignore [hunk 8]
```diff
@@ -1 +1,3 @@
 node_modules
+.env
+localhostconnect.js
```
Comment (leafoflegend) [inline .gitignore:3]: Probably isn't enough, but good to see `.env` made the list.

File: api/index.js [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+  const auth = req.header("Authorization");
+
+  if (!auth) {
+    next();
+  } else if (auth.startsWith(prefix)) {
+    const token = auth.slice(prefix.length);
+
+    try {
+      const { id } = jwt.verify(token, process.env.JWT_SECRET);
+      // console.log(jwt.verify(token, process.env.JWT_SECRET));
```
Comment (leafoflegend) [inline api/index.js:22]: What ah... value does this comment provide?

File: api/index.js [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+      if (id) {
+        req.user = await getUserById(id);
```
Comment (leafoflegend) [inline api/index.js:24]: Nice.

File: api/index.js [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+        next();
+      } else if (!id) {
+        next({ message: "Authorization is not working" });
```
Comment (leafoflegend) [inline api/index.js:27]: Bad message, its more like `"JWT Verification Failed."`.

File: api/orders.js [hunk 12]
```diff
@@ -0,0 +1,6 @@
+require("dotenv").config();
+const express = require("express");
+const router = express.Router();
+const { requireUser } = require("./utils.js");
+
+const { createOrder, getAllOrders, getOrderById } = require("../db");
```
Comment (leafoflegend) [inline api/orders.js]: Interesting file? Seems ready for express work?

File: api/products.js [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+
+function isAdmin(req, res, next) {
+  const user = req.user;
+
+  if(user && user.isAdmin) {
+    next()
+  }else{
+    res.status(403).json({message: 'Only admins can perform this action'})
+  }
+}
```
Comment (leafoflegend) [inline api/products.js:13]: v nice.

File: api/products.js [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+    res.send(products)
+  }catch(error){
+    next(error)
+  }
+})
+
+//delete product(only admin can delete products)
+router.delete("/:productId",isAdmin, async (req, res, next)=>{
+  try{
+    const productId = req.params.Id;
```
Comment (leafoflegend) [inline api/products.js:46]: This won't work. The param is `productId`.

File: api/users.js [hunk 15]
```diff
@@ -0,0 +1,139 @@
+require("dotenv").config();
```
Comment (leafoflegend) [inline api/users.js:1]: This shouldn't be here.

File: api/users.js [hunk 16]
```diff
@@ -0,0 +1,139 @@
...
+    if (password.length < passwordMinLength) {
+      //console.log("register_password", password);
+      res.send({
+        name: "PasswordMustBe8CharactersError",
+        message: "Password Too Short!",
+      });
+    } else {
+      const user = await createUser({
+        username,
+        password,
```
Comment (leafoflegend) [inline api/users.js:55]: Shouldn't we *b crypt*'in this password?

File: api/users.js [hunk 17]
```diff
@@ -0,0 +1,139 @@
...
+// GET /api/users/profile
+router.get("/:username/profile", requireUser, async (req, res, next) => {
+  const { username } = req.params;
+  try {
+    const currentUser = await getUserByUsername(username);
+    if (!currentUser) {
+      res.send({
+        name: "WrongProfileUser",
+        message: "You are not permitted to see this profile!",
+      });
```
Comment (leafoflegend) [inline api/users.js:128]: Does this work like you suggest here? It seems like one could submit any username and pass inspection?

File: app.js [hunk 18]
```diff
@@ -0,0 +1,20 @@
+require("dotenv").config();
+const express = require("express");
+const app = express();
+const apiRouter = require("./api");
+const morgan = require("morgan");
+const cors = require("cors");
+const bodyParser = require("body-parser");
+
+app.use(bodyParser.json());
```
Comment (leafoflegend) [inline app.js:9]: `app.use(express.json());` works and gets rid of a dependency.

File: db/users.js [hunk 19]
```diff
@@ -0,0 +1,110 @@
+const client = require("./client");
+const bcrypt = require("bcrypt");
+
+async function createUser({ username, password, email, isAdmin }) {
+  const SALT_COUNT = 10;
+  const hashedPassword = await bcrypt.hash(password, SALT_COUNT);
```
Comment (leafoflegend) [inline db/users.js:6]: Ah, ignore some earlier commentary. You *B CRYPT*'in!

File: package.json [hunk 20]
```diff
@@ -0,0 +1,41 @@
+{
+  "name": "tacos",
+  "version": "1.0.0",
+  "main": "index.js",
+  "scripts": {
+    "seed:dev": "nodemon server.js",
```
Comment (leafoflegend) [inline package.json:6]: The newest version of node just have `--watch` which is nice so you don't need this.

**Review by leafoflegend — CHANGES_REQUESTED** [review]
Left a few comments here. Things like fairly tight. Security looks mostly good, think the only vuln I caught was the profiles one.

Prisma is a nice touch if thats working? It looks like it may be mid-flight.

Given time remaining, you really need to get pushing on the front end.

File: api/products.js (hunk 1 above)
Comment (harshilr98) [reply api/products.js]: done

File: api/index.js (hunk 10 above)
Comment (harshilr98) [reply api/index.js:24]: :D

File: api/users.js (hunk 17 above)
Comment (Drewseph87) [reply api/users.js]: we believe resolved, set it to getUserbyUserID instead to focus on the current person

**Review by leafoflegend — APPROVED** [review]
🚢 it!

**Pull request merged**